### ISSUE TITLE
TOTP execution exception [NoSuchElementException: None.get]

### ISSUE THREAD

**Issue by bravegag** [mohiva/play-silhouette-seed#109]
This occurs at `totp.scala.html#l32` and it happens when the user enters an invalid or expired TOTP code (when the code is correct it works). The issue is due to the `TotpController.view` not populating the `TotpForm` "hidden fields" needed by the auth process. 

This issue was originally reported in the Silhouette forum:
http://discourse.silhouette.rocks/t/execution-error-in-silhouette-seed-template/370

A patch will follow.

**Comment by bravegag** [comment]
The fix corresponds to the following two files:

```
bravegag@Zeus:~/code/play-silhouette-seed$ gits
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   app/controllers/TotpController.scala
	modified:   conf/routes

no changes added to commit (use "git add" and/or "git commit -a")
```

and here goes the FIX content diff:

```
diff --git a/app/controllers/TotpController.scala b/app/controllers/TotpController.scala
index 6857441..9781df3 100644
--- a/app/controllers/TotpController.scala
+++ b/app/controllers/TotpController.scala
@@ -47,8 +47,8 @@ class TotpController @Inject() (
    * Views the `TOTP` page.
    * @return The result to display.
    */
-  def view = silhouette.UnsecuredAction.async { implicit request =>
-    Future.successful(Ok(views.html.totp(TotpForm.form)))
+  def view(userId: java.util.UUID, sharedKey: String, rememberMe: Boolean) = silhouette.UnsecuredAction.async { implicit request =>
+    Future.successful(Ok(views.html.totp(TotpForm.form.fill(TotpForm.Data(userId, sharedKey, rememberMe)))))
   }
 
   /**
@@ -94,7 +94,7 @@ class TotpController @Inject() (
           case _ => Future.successful(Redirect(routes.ApplicationController.index()).flashing("error" -> Messages("invalid.verification.code")))
         }.recover {
           case _: ProviderException =>
-            Redirect(routes.TotpController.view()).flashing("error" -> Messages("invalid.unexpected.totp"))
+            Redirect(routes.TotpController.view(user.userID, data.sharedKey, request.authenticator.cookieMaxAge.isDefined)).flashing("error" -> Messages("invalid.unexpected.totp"))
         }
       }
     )
@@ -108,14 +108,15 @@ class TotpController @Inject() (
     TotpForm.form.bindFromRequest.fold(
       form => Future.successful(BadRequest(views.html.totp(form))),
       data => {
+        val totpControllerRoute = routes.TotpController.view(data.userID, data.sharedKey, data.rememberMe)
         userService.retrieve(data.userID).flatMap {
           case Some(user) =>
             totpProvider.authenticate(data.sharedKey, data.verificationCode).flatMap {
               case Some(_) => authenticateUser(user, data.rememberMe)
-              case _ => Future.successful(Redirect(routes.TotpController.view()).flashing("error" -> Messages("invalid.verification.code")))
+              case _ => Future.successful(Redirect(totpControllerRoute).flashing("error" -> Messages("invalid.verification.code")))
             }.recover {
               case _: ProviderException =>
-                Redirect(routes.TotpController.view()).flashing("error" -> Messages("invalid.unexpected.totp"))
+                Redirect(totpControllerRoute).flashing("error" -> Messages("invalid.unexpected.totp"))
             }
           case None => Future.failed(new IdentityNotFoundException("Couldn't find user"))
         }
diff --git a/conf/routes b/conf/routes
index caf42d4..87db0e3 100644
--- a/conf/routes
+++ b/conf/routes
@@ -13,7 +13,7 @@ POST        /signUp                          controllers.SignUpController.submit
 GET         /signIn                          controllers.SignInController.view
 POST        /signIn                          controllers.SignInController.submit
 
-GET         /totp                            controllers.TotpController.view
+GET         /totp                            controllers.TotpController.view(userId: java.util.UUID, sharedKey: String, rememberMe: Boolean)
 GET         /enableTotp                      controllers.TotpController.enableTotp
 GET         /disableTotp                     controllers.TotpController.disableTotp
 POST        /totpSubmit                      controllers.TotpController.submit
```

I can't fork-PR because I forked this repo long time ago and have a lot of extensions that I don't want to lose. My bad.

**Comment by bravegag** [comment]
Here the zippped patch.

[patch.zip](https://github.com/mohiva/play-silhouette-seed/files/3664610/patch.zip)

**Comment by akkie** [comment]
> I can't fork-PR because I forked this repo long time ago and have a lot of extensions that I don't want to lose. My bad.

Normally you can use a branch in your clone that is based on the actual master of this repo. There is no need to reset your changes.

**Comment by akkie** [comment]
Fixed with #111
